Orange/data/table/base.py

```diff
@@ -0,0 +1,1292 @@
+
+        This only does shallow inference on data types. Example:
+        if given a numpy matrix of dtype object (e.g. mixed numbers and strings),
+        pandas will interpret all columns as objects, and so will we.
+
+        Where possible, column names are preserved form the input, otherwise they are named
+        "Feature <n>", "Class <n>", "Target <n>" or "Meta <n>".
+        The domain is marked as anonymous.
+
+        Parameters
+        ----------
+        X_or_data : np.ndarray | pd.DataFrame
+            Either the X component of the data as a numpy ndarray or a complete pandas DataFrame.
+            If this is the only data argument, the column roles are also inferred.
+        Y : np.ndarray, optional, default None
+            The Y component of the data. Always assigned to Y in the result.
+        meta : np.ndarray, optional, default None
+            The meta attributes of the data. Always assigned to metas in the result.
+        infer_roles : bool, optional, default True
+            Whether to enable column role inference.
+
+        Returns
+        -------
+        TableBase
+            Return a new Table with the inferred domain.
+        """
+        role_vars = {'x': [], 'y': [], 'meta': []}
+        result = cls()
+
+        # used for data inference, shape checks, consolidated access
+        X_df = pd.DataFrame(data=X_or_data)
+        Y_df = pd.DataFrame(data=Y)
+        meta_df = pd.DataFrame(data=meta)
+
+        # transform any values we believe are null into actual null values
+        X_df.replace(to_replace=list(Variable.MISSING_VALUES), value=np.nan, inplace=True)
+        Y_df.replace(to_replace=list(Variable.MISSING_VALUES), value=np.nan, inplace=True)
+        meta_df.replace(to_replace=list(Variable.MISSING_VALUES), value=np.nan, inplace=True)
+
+        # override, because the user wishes to specify roles manually
+        if Y is not None or meta is not None:
+            infer_roles = False
+
+        # process every input segment with its intended role
+        for df, initial_role in zip((X_df, Y_df, meta_df), ('x', 'y', 'meta')):
+            for column_name, column in ((c, df[c]) for c in df.columns):
+                t, r = Domain.infer_type_role(column, initial_role if not infer_roles else None)
+                name = Domain.infer_name(t, r, column_name,
+                                         role_vars['x'], role_vars['y'], role_vars['meta'])
+                if t is DiscreteVariable:
+                    var = t(name, values=DiscreteVariable.generate_unique_values(column))
+                else:
+                    var = t(name)
+                result[var.name] = column
+                role_vars[r].append(var)
+        result.domain = Domain(role_vars['x'], role_vars['y'], role_vars['meta'])
+        result.domain.anonymous = True
+        result._transform_discrete_into_categorical()
+        result._transform_timevariable_into_datetime()
```

Comment:
> these won't need to be changed any time soon

Your famous last words?

Sure we know what they do, but how do we ensure we don't forget to call one of them somewhere?